option_three/client.py

```diff
@@ -0,0 +1,119 @@
+import asyncio
+import binascii
+import itertools
+import logging
+from typing import Sequence
+
+import grpc
+from proto import scraping_pb2, scraping_pb2_async_grpc
+from pyquery import PyQuery as pq
+from aiofile import AIOFile
+
+SCRAPER_POOL = ['127.0.0.1:10000']
+ENTRY_PAGE = 'https://www.google.com'
+BATCH_SIZE = 5
+OUTPUT_FOLDER = 'current/data'
+
+
+class ScrapingState:
+    """Stateful objects for scraping."""
+
+    def __init__(self):
+        self._visited_links = set()
+        self._pending_links = []
+        # No longer needs lock
+
+    def has_pending_links(self) -> bool:
+        """Return True if there is any pending link."""
+        return len(self._pending_links) > 0
+
+    def get_pending_link(self) -> str:
+        """Get one pending link."""
+        return self._pending_links.pop(0)
+
+    def new_links(self, links: Sequence[str]):
+        """Insert new links, update pending links."""
+        new_links_set = set(links)
+        unique_new_links = new_links_set - self._visited_links
+        self._pending_links.extend(list(unique_new_links))
+        self._visited_links |= new_links_set
+        logging.info('Visited links [%d] pending links [%d]', len(
+            self._visited_links), len(self._pending_links))
+
+
+def strip_links(page_content: str) -> Sequence[str]:
+    """Parse links from HTML."""
+    elements = pq(page_content)('a')
+    return [elements.eq(i).attr['href'] for i in range(len(elements))]
+
+
+class ScraperController:
+
+    def __init__(self, scraper_url: str, state: ScrapingState):
+        logging.info('Connecting to scraper server [%s]', scraper_url)
+        # Create gRPC Channel
+        self._channel = grpc.insecure_channel(scraper_url)
+        self._stub = scraping_pb2_async_grpc.ScraperStub(self._channel)
```

Comment:
AsyncScraperStub?

Reply:
Please note the async package.
While writing I've become more convinced that it's cleaner to separate sync and async stubs into separate packages due to issues outlined in `Examples.md`, so "option_three" example sticks to it

Reply:
I personally don't like previous design of different invocation methods. I think the async API is our chance to unify them. My rational -> https://github.com/lidizheng/proposal/blob/grpc-python-async-api/L58-python-async-api.md#unified-stub-call

Reply:
How would "Unified Stub Call" look like in typed world? Can you provide an example of generated typing for stubs?

Reply:
To support strong typing, it requires a separate file for Python 3. And for Python 2, existing generated code stays the same.

In Python 2, without typing, it doesn't matter
```Python
class GreeterStub(object):
  def __init__(self, channel):
    self.SayHello = channel.unary_unary(
        '/helloworld.Greeter/SayHello',
        request_serializer=helloworld__pb2.HelloRequest.SerializeToString,
        response_deserializer=helloworld__pb2.HelloReply.FromString,
        )
```

In Python 3, with `typing`, we can restrict the type of channel. `self.SayHello` is an instance of `grpc.aio.UnaryUnaryMultiCallable` class.
```Python3
class GreeterStub(object):
  def __init__(self, channel: grpc.aio.Channel):
    self.SayHello = channel.unary_unary(
        '/helloworld.Greeter/SayHello',
        request_serializer=helloworld__pb2.HelloRequest.SerializeToString,
        response_deserializer=helloworld__pb2.HelloReply.FromString,
        )
```

Reply:
As I've mentioned in other threads mypy does support typing for both Python 2 and Python 3, so typing should be roughly the same.

What is a concrete type of `GreeterStub.SayHello`?
It doesn't look like `grpc.aio.UnaryUnaryMultiCallable` is typed in the pull request.
For instance, what methods does it have, what are return types of this methods, how do we ensure that we can't call it with wrong arguments, and so on.

Reply:
The current pattern of generated stub is awkward, it patches the class on the fly.

I just realized the `MultiCallable` pattern is bad at supporting strong types. The input and output types are generic proto messages. I think the stub generation should be redesigned.

Ideally, the function signature should look like:

```Python
class GreeterStub:
    def __init__(self, ...):
        pass

    async def SayHello(self,
                       request: helloworld_pb2.GreetingRequest,
                       timeout: int=None,
                       metadata: Sequence=None,
                       credentials: grpc.CallCredentials=None,
                       wait_for_ready: bool=None,
                       compression: grpc.Compression=None) -> helloworld_pb2.GreetingResponse:
        pass
```

The configuration variables in `__call__` signature can very, depends on which pattern we are going to use (keyword arguments, context class, `contextvars`).

I'm fine to bump up the Python version requirement to 3.7, after the adoption 3.7 is more common.

Reply:
Does it mean we're giving up on `Unified Stub Call`?

Reply:
Nope. That means the three call options still merge into one. In the doc, I defined the ideal usage. It is the implementation that needs to be changed.

Reply:
How would a typing look like for stubs after the merge?
The snippet provided above only addresses "simple" case without any control over the headers/trailers/lifecycle.

Reply:
In the doc, the usage snippet that contains metadata/lifecycle/error handling.

```Python
# Usage 1: A simple call
response = await stub.Hi(...)

# Usage 2: Check metadata
call = stub.Hi(...)
if validate(await call.initial_metadata()):
    response = await call
    print(f'Getting response [{response}] with code [{call.code()}]')
else:
    raise ValueError('Failed to validate initial metadata')

# Usage 3: Control the life cycle
call = stub.Hi(...)
await async_stuff_that_takes_time()
if call.is_active() and call.time_remaining() < REMAINING_TIME_THRESHOLD:
    call.cancel()

# Usage 4: Error handling
try:
    response = await stub.FailedHi(...)
except grpc.RpcError as rpc_error:
    print(f'RPC failed: {rpc_error.code()}')
```

Reply:
Yes, but doc doesn't talk about typing (at all), and it's unclear how should `stub.Hi`'s signature look like to cover all of this use cases.
Can you help to clarify it, please?

Reply:
I'm a little confused about what you are asking for, can you be more straight forward?

To what I understand, my original question here is:
1. Should we generate a Python 3 only file?
2. Should we inject "async" method into `MultiCallable` class? Or other mechanism that allows one stub object can call both sync / async API.

The question 1 is resolved.

My standing on question 2 is that existing `MultiCallable` is over-complicated, and all the methods can be merged into one. And the usage of new API fits `asyncio` context.

Then, for the question 2, I raised the question about strong typing in `MultiCallable` class. Since you are familiar with gRPC Python, you can understand why that might be a problem. So, in previous comment, I am proposing the direction removing the `MultiCallable` abstraction layer.

```Python
# Generated Async Stub
class GreeterAsyncStub:
    def __init__(self, ...):
        pass

    async def SayHello(self,
                       request: helloworld_pb2.GreetingRequest,
                       timeout: int=None,
                       metadata: Sequence=None,
                       credentials: grpc.CallCredentials=None,
                       wait_for_ready: bool=None,
                       compression: grpc.Compression=None) -> Future[helloworld_pb2.GreetingResponse]:
        ...
        return grpc.aio.Call(...)

# grpc.aio.Call class
class Call(asyncio.Task, grpc.Context):
    async def initial_metadata(self) -> Sequence:
        pass

```

Reply:
Sure. Let's consider the definition of `SayHello` from the snippet above:
```python
async def SayHello(self, ...) -> Future[helloworld_pb2.GreetingResponse]:
  pass

call = stub.SayHello(request)
call.initial_metadata()  # <- type failure, `call` does not have "initial_metadata" method
call.time_remaining()  # <- type failure, unknown method
```

I guess changing it to smth like this should help (note removing `async` and returning `Call` object instead of `asyncio.Future`):
```
class GreeterAsyncStub:
    def __init__(self, ...):
        pass

    def SayHello(self,
                       request: helloworld_pb2.GreetingRequest,
                       timeout: int=None,
                       metadata: Sequence=None,
                       credentials: grpc.CallCredentials=None,
                       wait_for_ready: bool=None,
                       compression: grpc.Compression=None) -> grpc.aio.Call[helloworld_pb2.GreetingResponse]:
        ...
        return grpc.aio.Call(...)

# grpc.aio.Call class
T = TypeVar('T')
class Call(typing.Awaitable[T], grpc.Context):
  pass
```

Reply:
Yes, return type looks good.

Reply:
Good point about this @euroelessar 

> .. note removing async and returning Call object instead of asyncio.Future

I'm just in the implementation of the unified call object and I need to change the method from an async one to sync one, otherwise, the `await` used by the caller will be applied to the coroutine instead of the `Call` object.

My main concern with this is the friction that it would imply with the interceptors, since then by nature would need to be coroutines but would need to be added between the caller and the final call to the stub, which would not work.

In any case, I'm working on it and I will try to come up with something that works.